src/Middleware/RateLimiting/src/RateLimitingMiddleware.cs

```diff
@@ -14,26 +15,49 @@ namespace Microsoft.AspNetCore.RateLimiting;
 internal sealed partial class RateLimitingMiddleware
 {
     private readonly RequestDelegate _next;
-    private readonly Func<HttpContext, RateLimitLease, Task> _onRejected;
+    private readonly Func<OnRejectedContext, CancellationToken, ValueTask>? _onRejected;
     private readonly ILogger _logger;
-    private readonly PartitionedRateLimiter<HttpContext> _limiter;
+    private readonly PartitionedRateLimiter<HttpContext>? _globalLimiter;
+    private readonly PartitionedRateLimiter<HttpContext> _endpointLimiter;
     private readonly int _rejectionStatusCode;
+    private readonly IDictionary<string, DefaultRateLimiterPolicy> _policyMap;
+    private readonly DefaultKeyType _defaultPolicyKey = new DefaultKeyType<PolicyNameKey>(new PolicyNameKey { PolicyName = "__defaultPolicyKey" });
+    private readonly IServiceProvider _serviceProvider;
 
     /// <summary>
     /// Creates a new <see cref="RateLimitingMiddleware"/>.
     /// </summary>
     /// <param name="next">The <see cref="RequestDelegate"/> representing the next middleware in the pipeline.</param>
     /// <param name="logger">The <see cref="ILogger"/> used for logging.</param>
     /// <param name="options">The options for the middleware.</param>
-    public RateLimitingMiddleware(RequestDelegate next, ILogger<RateLimitingMiddleware> logger, IOptions<RateLimiterOptions> options)
+    /// <param name="serviceProvider">The service provider for this middleware.</param>
+    public RateLimitingMiddleware(RequestDelegate next, ILogger<RateLimitingMiddleware> logger, IOptions<RateLimiterOptions> options, IServiceProvider serviceProvider)
     {
         _next = next ?? throw new ArgumentNullException(nameof(next));
 
         _logger = logger ?? throw new ArgumentNullException(nameof(logger));
 
-        _limiter = options.Value.Limiter;
+        _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
+
         _onRejected = options.Value.OnRejected;
-        _rejectionStatusCode = options.Value.DefaultRejectionStatusCode;
+        _rejectionStatusCode = options.Value.RejectionStatusCode;
+        _policyMap = options.Value.PolicyMap;
+
+        // Use reflection to activate policies passed to AddPolicy<TPartitionKey, TPolicy>
+        var convertPolicyObject = typeof(RateLimiterOptions).GetMethod("ConvertPolicyObject");
+
+        foreach (var policyTypeInfo in options.Value.UnactivatedPolicyMap)
+        {
+            var genericConvertPolicyObject = convertPolicyObject!.MakeGenericMethod(policyTypeInfo.Value.PartitionKeyType);
+            var instance = ActivatorUtilities.CreateInstance(_serviceProvider, policyTypeInfo.Value.PolicyType);
+            var obj = genericConvertPolicyObject.Invoke(new RateLimiterOptions(), new object[] { instance });
```

Comment:
Probably should be able to do all this without reflection, but it's Friday, so I'll take another look on Monday

Reply:
So there's an IOptions RateLimiterOptions which has the policy map, and we also use this same type as instances for policies?

I think it would be clearer to have two separate types, one for the IOptions, and another which doesn't go into DI, called RateLimiterPartitionSettings or whatever, to disambiguate the things that are accessed via Options vs instance data you are passing to these func/partition thingies (not sure what the correct name for obj) is